### ISSUE TITLE
Default entryPoints to shader modules

### ISSUE THREAD

**Issue by greggman** [gpuweb/gpuweb#4342]
Maybe this was already discussed but, as it is, you always have to specify entry points when creating a pipeline. After having written a bunch of WebGPU, it seems like it would be nice if there were some defaults. 

In other words, these 2 lines

```js
  const pipeline = device.createRenderPipeline({
    layout: 'auto',
    vertex: {
      module: shaderModule,
      entryPoint: 'myVSMain',    // <-------------------
    },
    fragment: {
      module: shaderModule,
      entryPoint: 'myFSMain',    // <-------------------
      ...
    },
  });
```

If there were a default "entryPoint" for each stage then these lines would not be needed and would make the api a tiny bit less tedious.

4 ideas

1. The defaults could be something like `vertexMain`, `fragmentMain`, `computeMain`
2. If the attribute is considered part of the name then just `main`

   In other words, `@compute fn main(...)` is different from `@vertex main fn(...)` because
   of the attribute specifying what type of shader it is so if in createXXXPipeline you don't specify an entry point then it looks for `main` with a correct attribute for the stage.

3. The default is the first compatible function.

   In this idea, using the attributes again `@compute` vs `@fragment` vs `@vertex`. If no entryPoint is specified, find the first function that has the required attribute, the name is irrelevant.

4. The default is the only compatible function

   This idea is the same as # 3 except if there are 2 or more compatible functions you get an error "More than one compatible function. You must specify an entryPoint"

I know it's not a big deal. I just thought I ask.

**Comment by Kangz** [comment]
Option 3 seems pretty good IMHO.

**Comment by ben-clayton** [comment]
I would prefer Option 4 over Option 3, as WGSL's module-scope declarations are order independent. Option 3 introduces ordering again.

I also struggle to think of a good reason _why_ you'd want two or more entry points with the same stage in the same shader, especially if the first is only ever picked. Sounds like a bug, which 3 hides and 4 alerts you to.

**Comment by ben-clayton** [comment]
> I also struggle to think of a good reason _why_ you'd want two or more entry points with the same stage in the same shader, especially if the first is only ever picked. Sounds like a bug, which 3 hides and 4 alerts you to.

After posting, I realise a use case would be mixing 'default' entry points with explicitly named ones. Sounds messy.

**Comment by Kangz** [comment]
Sorry I meant option 4), for the reasons that @ben-clayton exposed.

**Comment by kainino0x** [comment]
Option 4 SGTM (it was going to be my suggestion before I even finished reading the post)

**Comment by dneto0** [comment]
Option 4 SGTM.

**Comment by dneto0** [comment]
Set to milestone 2 for triage

**Comment by kainino0x** [comment]
Just a note: this won't apply to the entryPoint in the GPUShaderModuleCompilationHint, because we rely on it to know which shader stage you're referring to.
https://gpuweb.github.io/gpuweb/#dom-gpushadermodulecompilationhint-entrypoint

**Comment by floooh** [comment]
> I also struggle to think of a good reason why you'd want two or more entry points with the same stage in the same shader

A bit tangential (because entirely different shading language and different library), but in my sample code I often put all required shaders into the same source file, and often also share structs and helper functions between "unrelated" shaders.

For instance this has two sets of shaders (one for an offscreen render pass, and one for the default render pass), and shares a uniform block struct between two vertex shaders.

https://github.com/floooh/sokol-samples/blob/master/sapp/offscreen-sapp.glsl

...in straight WGSL this would be a lot more straightforward because all those non-standard "meta-tags" wouldn't be needed.

**Comment by Kangz** [comment]
Maybe we can resolve this offline. @teoxoy @jimblandy @mwyrzykowski, would you be ok to go with option 4)? This is a nice quality of life improvement for developers.

**Comment by mwyrzykowski** [comment]
Option (4) please but I would say we leave in M2 as tagged

**Comment by teoxoy** [comment]
The 4th option sounds like the best one to me as well.

**Comment by jimblandy** [comment]
I like option 4.

But 'compatible' should only refer to the stage attributes: if there's only one function with the `@vertex` attribute in the module, say, picking that one seems fine and good. But we should definitely not take, say, vertex shader arguments into account when deciding whether there's only one "suitable" entry point.

I think that's what was intended, but I wanted to make sure.

**Comment by Kangz** [comment]
Consensus in [the last meeting](https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2023-11-08) was to go with option 4 (and the simple validation @jimblandy mentioned).
